src/core/api.rs

```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
+    /// policy.
+    ///
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encaps(
+        &self,
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
+    ) -> Result<(Secret<SEED_LENGTH>, Encapsulation), Error> {
+        let ap = &AccessPolicy::parse(ap)?;
         encaps(
             &mut *self.rng.lock().expect("Mutex lock failed!"),
-            pk,
-            &policy.generate_point_coordinates(access_policy)?,
+            mpk,
+            &policy.generate_point_coordinates(ap.clone())?,
         )
     }
-
-    /// Attempts opening the given `Covercrypt` encapsulation. Returns the
-    /// encapsulated key upon success and `None` otherwise.
-    // TODO document error cases.
-    pub fn decaps(
+    /// Attempts opening the given encapsulation using the given
+    /// user secret key.
+    ///
+    /// Returns the encapsulated symmetric key if the user key holds
+    /// the correct rights.
+    fn decaps(
         &self,
         usk: &UserSecretKey,
-        encapsulation: &Encapsulation,
-    ) -> Result<Option<SymmetricKey<SEED_LENGTH>>, Error> {
-        decaps(usk, encapsulation)
+        encapsulation: Encapsulation,
+    ) -> Option<Secret<SEED_LENGTH>> {
+        match decaps(usk, &encapsulation) {
+            Err(_) => None,
+            Ok(sk) => Some(sk?),
+        }
     }
+}
 
-    /// Encrypts the given plaintext using the given symmetric key.
+pub trait CovercryptPKE<
+    Aead,
+    const KEY_LENGTH: usize,
+    const NONCE_LENGTH: usize,
+    const MAC_LENGTH: usize,
+>
+{
+    /// Encrypts the given plaintext using Covercrypt and the given DEM.
     ///
-    /// The encryption scheme used is AES-256 GCM.
+    /// Creates a Covercrypt encapsulation of a LENGTH-byte key, and use this
+    /// key with the given authentication data to produce a DEM ciphertext of
+    /// the plaintext.
     ///
-    /// - `symmetric_key`   : AES key
-    /// - `plaintext`       : data to be encrypted
-    /// - `ad`              : optional associated data
-    pub fn dem_encrypt(
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn encrypt(
         &self,
-        symmetric_key: &SymmetricKey<SEED_LENGTH>,
-        plaintext: &[u8],
+        mpk: &MasterPublicKey,
+        policy: &Policy,
+        ap: &str,
         ad: Option<&[u8]>,
-    ) -> Result<Vec<u8>, Error> {
-        let aes256gcm = Aes256Gcm::new(symmetric_key);
-        let nonce = Nonce::new(&mut *self.rng.lock().expect("could not lock mutex"));
-        let mut ciphertext = aes256gcm.encrypt(&nonce, plaintext, ad)?;
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
-        res.append(&mut ciphertext);
-        Ok(res)
-    }
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error>;
```

Comment:
I wonder if we should already serialize the encapsulation: the current solution seems to be somewhere in-between returning a struct containing each parts (encapsulation, mac, none and ciphertext) and serializing and concatenating all parts.

I don't have any strong opinion on it but this solution is surprising.